compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp

```diff
@@ -309,6 +309,20 @@ class ConvertIREEBindingSubspanOp : public ConvertToLLVMPattern {
     llvmFuncOp.setArgAttr(llvmBufferArg.getArgNumber(),
                           LLVM::LLVMDialect::getAlignAttrName(),
                           rewriter.getI32IntegerAttr(16));
+    // It is safe to set the noalias attribute as it is guaranteed that the
+    // ranges within bindings won't alias.
+    llvmFuncOp.setArgAttr(llvmBufferArg.getArgNumber(),
+                          LLVM::LLVMDialect::getNoAliasAttrName(),
+                          rewriter.getUnitAttr());
+    if (bitEnumContainsAny(subspanOp.getDescriptorFlags().value_or(
+                               IREE::HAL::DescriptorFlags::None),
+                           IREE::HAL::DescriptorFlags::ReadOnly)) {
+      // Setting the readonly attribute here will generate non-coherent cache
+      // loads.
+      llvmFuncOp.setArgAttr(llvmBufferArg.getArgNumber(),
```

Comment:
There may be several BindingSubspanOp matching to the same kernel argument. How do we know that they all have the same matching flag?
@benvanik is it safe to make this assumption? 

If it is safe I would at least verify that assumption somewhere and making sure the IR matches it.

Reply:
(scoping to the actual `llvmBufferBasePtr` will be much better than the workaround and is worth figuring out if we can)

Reply:
Unfortunately this flag can only be set on function arguments. Creating a shim function wouldn't help as we would need to inline and the information would be lost for the backend. It may help some other optimizations.
An alternative is to set the read_only marker directly onto the load but this would most likely require more work to plumb that through.

> A workaround here would be to only add the readonly attribute if all bindings mapping to the kernel arg have the flag. That's a bit tricky to tell from a conversion pattern in here, though. Instead the main function conversion pattern that defines the kernel args could walk itself to find all subspan ops, AND their flags together, and produce the kernel args with the right attributes. This pattern would ignore the flag and just inherit what was already set.

That makes sense and it is most likely the only solution feasible with the existing flow.

Reply:
I think we're hitting the limits of this current implementation while trying to enable `--iree-stream-resource-alias-mutable-bindings=true` by default (https://github.com/openxla/iree/pull/14658). That flag shows promising results (10-15% latency improvements) on other backends and is required for correctness on WebGPU, but aliasing the constant and mutable bindings gets the CUDA target off of this fast path of `readonly` + `noalias`.

For example, without the flag, we can have IR like this before `ConvertToNVVM`:
```mlir
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4348864) flags(ReadOnly) : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    memref.assume_alignment %0, 64 : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c27388864) : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
    memref.assume_alignment %1, 64 : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
```

and with the flag, the bindings are aliased (`binding(1)` to `binding(0)`):
```mlir
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4348864) : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    memref.assume_alignment %0, 64 : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    %1 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c27388864) : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
    memref.assume_alignment %1, 64 : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
```

Then after `ConvertToNVVM`, the signature changes from
```mlir
  llvm.func @forward_dispatch_2(%arg0: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias, llvm.readonly}, %arg1: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias}) {
```

to
```mlir
llvm.func @forward_dispatch_2(%arg0: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias}) {
```
and none of the pointer loads keep the `readonly` attribute

> For future work, is it possible to cluster read-only subspanOps in the same kernel argument?
>
> Another idea is to mark the load as read-only. But here I should convey the region information of each subspanOp to llvm dialect and work with the pointers there. This is not trivial.

Are either of these approaches viable?

Reply:
What is the motivation of enabling aliasing for Nvidia GPUs?

Reply:
> What is the motivation of enabling aliasing for Nvidia GPUs?

My feeling is that since we set up the resource bindings in `stream` (prior to `hal`), we'd prefer to construct them in a target-agnostic way. We would then expect each target (CPU, Vulkan with some target triple, CUDA, etc.) to be able to lower to efficient code. As it stands now, we have some target/device configurations that appear to benefit from aliasing (according to our benchmarks) and some that don't.